File: content/LLD/Reference-Code/22. Ride Sharing Project - Good Example/ride.py
from enum import Enum
from passenger import Passenger
from driver import Driver
from fare_strategy import FareStrategy


class RideStatus(Enum):
    SCHEDULED = "SCHEDULED"
    ONGOING = "ONGOING"
    COMPLETED = "COMPLETED"


class Ride:
    def __init__(
        self,
        passenger: "Passenger",
        driver: "Driver",
        distance: float,
        fare_strategy: "FareStrategy",
    ):
        self.passenger = passenger
        self.driver = driver
        self.distance = distance
        self.fare_strategy = fare_strategy
        self.fare: float = 0.0
        self.status: RideStatus = RideStatus.SCHEDULED

    def calculateFare(self):
        self.fare = self.fare_strategy.calFare(self.driver.get_vehicle(), self.distance)

    def getRideFare(self):
        return self.fare

    def updateStatus(self, new_ride_status: RideStatus):
        self.status = new_ride_status
        self.__notifyUsers(self.status)

    def __notifyUsers(self, ride_status: RideStatus):
        self.driver.notify(f"Your ride is {ride_status.value}")
        self.passenger.notify(f"Your ride status is {ride_status.value}")

File: content/LLD/Reference-Code/22. Ride Sharing Project - Good Example/Ride.java
public class Ride {
    public Passenger passenger;
    public Driver driver;
    public double distance;
    public FareStrategy fareStrategy;
    public double fare = 0.0;
    public RideStatus status = RideStatus.SCHEDULED;

    public Ride(Passenger passenger, Driver driver, double distance, FareStrategy fareStrategy) {
        this.passenger = passenger;
        this.driver = driver;
        this.distance = distance;
        this.fareStrategy = fareStrategy;
    }

    public void calculateFare() {
        this.fare = this.fareStrategy.calFare(this.driver.getVehicle(), this.distance);
    }

    public double getRideFare() {
        return this.fare;
    }

    public void updateStatus(RideStatus newRideStatus) {
        this.status = newRideStatus;
        notifyUsers(this.status);
    }

    private void notifyUsers(RideStatus rideStatus) {
        this.driver.notify("Your ride is " + rideStatus.getValue());
        this.passenger.notify("Your ride status is " + rideStatus.getValue());
    }
}
